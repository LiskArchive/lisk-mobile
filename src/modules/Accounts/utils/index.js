import moment from 'moment';

export const getPendingTime = (unvoteHeight, unlockHeight) => {
  const awaitingBlocks = unlockHeight - unvoteHeight;
  const secondsToUnlockAllBalance = awaitingBlocks * 10;
  const momentSeconds = moment().second(secondsToUnlockAllBalance);
  return moment().to(momentSeconds, true);
};
