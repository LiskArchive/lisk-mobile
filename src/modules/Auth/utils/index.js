export * from './passphrase';
export * from './download';
