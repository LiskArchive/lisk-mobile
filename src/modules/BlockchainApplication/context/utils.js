/**
 * Applications reducer function for the Blockchain Applications context.
 * @param {Array} state - Current state of applications.
 * @param {Object} action - Action to dispatch for updating the state.
 * Can be "init" applications, "add" an application or "delete" an application.
 * @returns {Array} state - The context state of saved applications.
 */
export function applicationsContextReducer(
  state = [],
  { type, applications, application, chainID }
) {
  switch (type) {
    case 'init':
      return applications;

    case 'add': {
      if (state?.find((app) => app.chainID === application.chainID)) {
        return state;
      }

      return [...state, application];
    }

    case 'delete': {
      return state.filter((app) => app.chainID !== chainID);
    }

    default:
      return state;
  }
}

/**
 * Pins reducer function for the Blockchain Applications context.
 * @param {Array} state - Current state of pins.
 * @param {Object} action - Action to dispatch for updating the state.
 * Can be "init" pins, "add" a pin or "delete" a pin.
 * @returns {Array} state - The context state of saved pins (chainIDs).
 */
export function applicationPinsContextReducer(state, { type, pins, chainID }) {
  switch (type) {
    case 'init':
      return pins;

    case 'add': {
      if (state.includes(chainID)) {
        return state;
      }

      return [...state, chainID];
    }

    case 'delete': {
      return state.filter((_chainID) => _chainID !== chainID);
    }

    default:
      return state;
  }
}

/**
 * Gets the applications array to provide to ApplicationsContext on app mount.
 * @param {Array} applications - All available applications data (fetched from the server).
 * @param {Array} defaultApplications - All default applications data (fetched from the server).
 * @param {Array} cachedChainIDs - Saved on local storage by the user applications chainIDs.
 * @returns {Array} Applications to insert on the ApplicationsContext first value.
 */
export function getInitContextApplications({
  applications,
  defaultApplications,
  cachedChainIDs = [],
}) {
  let cachedApplicationsQueue = applications.filter((app) => cachedChainIDs.includes(app.chainID));

  let initApplications = defaultApplications.reduce((acc, defaultApplication) => {
    const cachedApplication = cachedApplicationsQueue.find(
      (cachedApplicationInQueue) => cachedApplicationInQueue.chainID === defaultApplication.chainID
    );

    if (cachedApplication) {
      cachedApplicationsQueue = cachedApplicationsQueue.filter(
        (cachedApplicationInQueue) => cachedApplicationInQueue.chainID !== cachedApplication.chainID
      );

      if (!defaultApplications.find((accApp) => accApp.chainID === cachedApplication.chainID)) {
        return [...acc, { ...cachedApplication, ...defaultApplication }];
      }
    }

    return [...acc, defaultApplication];
  }, []);

  if (cachedApplicationsQueue.length > 0) {
    initApplications = initApplications.concat(cachedApplicationsQueue);
  }

  return initApplications;
}
