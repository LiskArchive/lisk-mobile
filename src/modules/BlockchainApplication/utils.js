import liskAPIClient from 'utilities/api/LiskAPIClient';
import { getApplicationsQueryConfigCreator } from './api/useApplicationsQuery';

/**
 * Merges on-chain and off-chain applications data into a single array of objects
 * by intercepting GET_APPLICATIONS_QUERY with GET_APPLICATIONS_META_QUERY before.
 * @param {Function} createQueryConfig - Function to create the GET_APPLICATIONS_META_QUERY
 * based on GET_APPLICATIONS_QUERY results.
 * @param {Object} apiClient - API client instance (default: liskAPIClient).
 */
function getApplicationsQueryTransformer({ createQueryConfig, apiClient }) {
  return async (applicationsMeta) => {
    try {
      const chainIDs = applicationsMeta?.map(({ chainID }) => chainID).join(',');

      if (!chainIDs) {
        return applicationsMeta;
      }

      const metaConfig = createQueryConfig({
        params: {
          chainID: chainIDs,
          limit: applicationsMeta.length,
        },
      });

      const applicationsData = await apiClient.call(metaConfig);

      return applicationsMeta.map((appMeta) => {
        const selectedAppData = applicationsData?.data?.find(
          (appData) => appData.chainID === appMeta.chainID
        );

        return { ...(selectedAppData ?? {}), ...appMeta };
      });
    } catch (error) {
      return applicationsMeta;
    }
  };
}

/**
 * Injects to GET_APPLICATIONS_QUERY chainID params the result apps fetched from
 * GET_APPLICATIONS_META_QUERY.
 * @param {Object} res - GET_APPLICATIONS_META_QUERY result.
 * @returns - Merged apps resulting from GET_APPLICATIONS_META_QUERY and
 * GET_APPLICATIONS_QUERY.
 */
export async function transformApplicationsMetaQueryResult(res) {
  const transformApplications = getApplicationsQueryTransformer({
    createQueryConfig: getApplicationsQueryConfigCreator(),
    apiClient: liskAPIClient,
  });

  const applicationsData = await transformApplications(res.data);

  return {
    ...res,
    data: applicationsData,
  };
}

/**
 * Calculates current application data by merging applications of-chain
 * and on-chain data.
 * @param {Object} applicationsMetadata - Applications off-chain data.
 * @param {Object} applicationsData - Applications on-chain data.
 * @returns - Current application data.
 */
export function getCurrentApplicationData(applicationsMetadata, applicationsData) {
  const applicationMetadata = applicationsMetadata[0];

  const applicationData = applicationsData.find(
    (appData) => appData.chainID === applicationMetadata.chainID
  );

  return { ...applicationMetadata, ...applicationData };
}
