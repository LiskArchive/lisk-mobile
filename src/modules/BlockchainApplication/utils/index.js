/**
 * Access next and previous items of an array in a round form
 * If current item is the last item of the array, next item would be the first item;
 * Also, if current item is the first item, previous would be the last item
 * @param {Array} arr - Array of data
 * @param {Number} index - Current index of current active data
 * @param {'next' | 'prev'} direction - Direction to move pointer for next data
 * @returns {any} - arr[nextIndex]
 */
export const roundAccessor = (arr = [], index = 0, direction = 'next') => {
  if (direction === 'next') {
    return arr[index + 1] ? arr[index + 1] : arr[0];
  }
  return index === 0 ? arr[arr.length - 1] : arr[index - 1];
};

/**
 * Merges on-chain and off-chain applications data into a single array of objects.
 * @param {Object} applicationsData
 * @param {Object} applicationsMetaData
 */
export function mergeApplicationsData(applicationsData, applicationsMetaData) {
  return applicationsData.map((appMetadata) => {
    const app = applicationsMetaData.find((_app) => _app.chainID === appMetadata.chainID);

    return { ...appMetadata, ...app };
  });
}
