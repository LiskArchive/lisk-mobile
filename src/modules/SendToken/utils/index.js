/* eslint-disable max-statements, max-len */
import { Platform } from 'react-native';
import { transactions, cryptography } from '@liskhq/lisk-client';
import * as transactionsConstants from 'modules/SendToken/constants';
import { fromRawLsk, toRawLsk } from 'utilities/conversions';
import computeMinFee from './fees';

export const createTransactionObject = (nonce, amount = 0, message = '', publicKey) => ({
  moduleID: cryptography.utils.intToBuffer('2', 4),
  commandID: cryptography.utils.intToBuffer('0', 4),
  // eslint-disable-next-line no-undef
  nonce: BigInt(nonce),
  // eslint-disable-next-line no-undef
  baseFees: BigInt(0),
  params: {
    // eslint-disable-next-line no-undef
    amount: BigInt(toRawLsk(Number(amount))),
    recipientAddress: Buffer.from(''),
    data: message,
  },
  signatures: [],
  senderPublicKey: Buffer.from(publicKey, 'hex'),
});

export const getTransactionFee = ({
  account,
  transaction,
  selectedPriority,
  selectedPriorityIndex,
}) => {
  try {
    const feePerByte = selectedPriority.fee ?? 0;
    const schema = transactionsConstants.transferAssetSchema;
    const maxAssetFee =
      transactionsConstants.moduleAssetMap[transactionsConstants.moduleCommandNameIdMap.transfer]
        .maxFee;

    const transactionObject = createTransactionObject(
      account.nonce,
      transaction.amount,
      transaction.data,
      account.publicKey
    );
    let minFee;
    if (Platform.OS === 'android') {
      minFee = computeMinFee(transactionObject, {
        baseFees: transactionsConstants.BASE_FEES,
      });
    } else {
      minFee = transactions.computeMinFee(transactionObject, schema, {
        baseFees: transactionsConstants.BASE_FEES,
      });
    }
    const tieBreaker =
      selectedPriorityIndex === 0
        ? 0
        : transactionsConstants.MIN_FEE_PER_BYTE * feePerByte * Math.random();
    const size = transactions.getBytes(transactionObject, schema).length;
    const calculatedFee = Number(minFee) + size * feePerByte + tieBreaker;
    const cappedFee = Math.min(calculatedFee, maxAssetFee);
    const feeInLsk = fromRawLsk(cappedFee.toString());
    const roundedValue = Number(feeInLsk).toFixed(7).toString();

    const feedback = transaction.amount === '' ? '-' : `${roundedValue ? '' : 'Invalid amount'}`;
    return {
      value: roundedValue,
      error: !!feedback,
      feedback,
    };
  } catch (error) {
    return { value: 0, error: false, feedback: '' };
  }
};
