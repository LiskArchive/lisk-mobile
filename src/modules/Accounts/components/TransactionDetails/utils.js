import { isRegistration, isTransfer, isVote } from 'modules/SendToken/constants';
import { stringShortener } from 'utilities/helpers';

export const goToWallet = (address, { navigation, account }) => {
  if (
    address !== account.address
    && address !== 'Unparsed Address'
  ) {
    navigation.navigate({ key: address, name: 'Wallet', params: { address } });
  }
};

export const getAccountLabel = (
  address,
  {
    t, followedAccounts, activeToken, truncate
  }
) => {
  if (address === 'Unparsed Address') {
    return t('Unparsed Address');
  }

  const followedAccount = followedAccounts[activeToken]?.find(
    a => a.address === address
  );
  if (followedAccount) {
    return followedAccount.label;
  }
  if (truncate) {
    return stringShortener(address, 6, 5);
  }

  return address;
};

export const getAccountTitle = tx => {
  if (isVote(tx)) return 'Voter';
  if (isRegistration(tx)) return 'Registrant';
  if (!isTransfer(tx) || tx.recipientAddress === tx.senderAddress) { return 'Account address'; }
  return 'Sender';
};
