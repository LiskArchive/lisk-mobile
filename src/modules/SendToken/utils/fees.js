import { getBytes } from 'utilities/api/lisk/utils';

const DEFAULT_MIN_FEE_PER_BYTE = 1000;
const DEFAULT_NUMBER_OF_SIGNATURES = 1;
const DEFAULT_BASE_FEE = '0';
const DEFAULT_SIGNATURE_BYTE_SIZE = 64;

const computeTransactionMinFee = (trx, options) => {
  const mockSignatures = new Array(
    options?.numberOfSignatures ?? DEFAULT_NUMBER_OF_SIGNATURES
  ).fill(Buffer.alloc(DEFAULT_SIGNATURE_BYTE_SIZE));
  if (options?.numberOfEmptySignatures) {
    mockSignatures.push(...new Array(options.numberOfEmptySignatures).fill(Buffer.alloc(0)));
  }
  const size = getBytes({
    ...trx,
    signatures: mockSignatures,
  }).length;
  const baseFee =
    options?.baseFees?.find((bf) => bf.moduleID === trx.moduleID && bf.commandID === trx.assetID)
      ?.baseFee ?? DEFAULT_BASE_FEE;
  // eslint-disable-next-line no-undef
  return BigInt(size * (options?.minFeePerByte ?? DEFAULT_MIN_FEE_PER_BYTE)) + BigInt(baseFee);
};

const computeMinFee = (trx, options) => {
  // eslint-disable-next-line no-undef
  trx.fee = BigInt(0);
  let minFee = computeTransactionMinFee(trx, options);

  // eslint-disable-next-line no-undef
  while (minFee > BigInt(trx.fee)) {
    // eslint-disable-next-line no-param-reassign
    trx.fee = minFee;
    minFee = computeTransactionMinFee(trx, options);
  }
  return minFee;
};

export default computeMinFee;
