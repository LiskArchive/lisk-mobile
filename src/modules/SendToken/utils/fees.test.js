import * as Lisk from '@liskhq/lisk-client';
import { getBytes } from 'utilities/api/lisk/utils';
import computeMinFee from './fees';

describe.skip('fee', () => {
  const passphrase1 = 'trim elegant oven term access apple obtain error grain excite lawn neck';
  const { publicKey: publicKey1 } = Lisk.cryptography.address
    .getAddressAndPublicKeyFromPassphrase(passphrase1);
  const validTransaction = {
    moduleID: 2,
    assetID: 0,
    // eslint-disable-next-line no-undef
    nonce: BigInt('1'),
    senderPublicKey: publicKey1,
    asset: {
      recipientAddress: Buffer.from('3a971fd02b4a07fc20aad1936d3cb1d263b96e0f', 'hex'),
      // eslint-disable-next-line no-undef
      amount: BigInt('4008489300000000'),
      data: '',
    },
  };
  const baseFees = [
    {
      moduleID: 2,
      assetID: 0,
      baseFee: '10000000',
    },
    {
      moduleID: 5,
      assetID: 0,
      baseFee: '1',
    },
    {
      moduleID: 3,
      assetID: 0,
      baseFee: '1',
    },
  ];

  describe('computeMinFee', () => {
    it('should return minimum fee required to send to network', () => {
      // Arrange
      const minFee = computeMinFee(validTransaction);

      // Assert
      expect(minFee).not.toBeUndefined();
      expect(minFee).toMatchSnapshot();
    });

    it('should calculate minimum fee for given minFeePerByte', () => {
      // Arrange
      const options = { minFeePerByte: 2000, baseFees, numberOfSignatures: 1 };
      const minFee = computeMinFee(validTransaction, options);

      // Assert
      expect(minFee).not.toBeUndefined();
      expect(minFee).toMatchSnapshot();
    });

    it('should calculate minimum fee for transaction from multisignature account', () => {
      // Arrange
      const options = { minFeePerByte: 2000, baseFees, numberOfSignatures: 64 };
      const minFee = computeMinFee(validTransaction, options);

      // Assert
      expect(minFee).not.toBeUndefined();
      expect(minFee).toMatchSnapshot();
    });

    it('should calculate minimum fee for transaction from multisignature account which has lower number of signatures than registered public keys', () => {
      // Arrange
      const options = {
        minFeePerByte: 1000,
        baseFees: [],
        numberOfSignatures: 2,
        numberOfEmptySignatures: 3,
      };
      const transaction = {
        ...validTransaction,
        signatures: [
          Buffer.alloc(64),
          Buffer.alloc(0),
          Buffer.alloc(0),
          Buffer.alloc(0),
          Buffer.alloc(64),
        ],
      };
      const minFee = computeMinFee(transaction, options);
      const txBytes = getBytes({ ...transaction, fee: minFee });

      // Assert
      // eslint-disable-next-line no-undef
      expect(minFee.toString()).toEqual(BigInt(txBytes.length * 1000).toString());
    });
  });
});
