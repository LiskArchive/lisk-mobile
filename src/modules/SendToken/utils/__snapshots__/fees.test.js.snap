// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fee computeMinFee should calculate minimum fee for given minFeePerByte 1`] = `290000n`;

exports[`fee computeMinFee should calculate minimum fee for transaction from multisignature account 1`] = `8608000n`;

exports[`fee computeMinFee should return minimum fee required to send to network 1`] = `145000n`;
